vumi/middleware/session_length.py

```diff
@@ -42,37 +42,43 @@ def setup_middleware(self):
     def teardown_middleware(self):
         yield self.redis.close_manager()
 
+    def _generate_redis_key(self, message, address):
+        return '%s:%s:%s' % (
+            message.get('transport_name'), address, 'session_created')
+
+    @inlineCallbacks
+    def _set_redis_time(self, redis_key, time):
+        if (yield self.redis.get(redis_key)) is None:
+            yield self.redis.setex(redis_key,  self.timeout, str(time))
+
     def _set_message_session_metadata(self, message, field, value):
-        if not message['helper_metadata'].get(self.field_name):
-            message['helper_metadata'][self.field_name] = {}
-        message['helper_metadata'][self.field_name][field] = value
+        metadata = message['helper_metadata'].setdefault(self.field_name, {})
+        metadata.setdefault(field, value)
 
     def _set_session_start_time(self, message, time):
+        time = float(time)
         self._set_message_session_metadata(message, 'session_start', time)
 
     def _set_session_end_time(self, message, time):
+        time = float(time)
         self._set_message_session_metadata(message, 'session_end', time)
 
     @inlineCallbacks
     def _process_message(self, message, redis_key):
         if message.get('session_event') == self.SESSION_NEW:
             start_time = self.clock.seconds()
-            yield self.redis.setex(redis_key,  self.timeout, str(start_time))
+            yield self._set_redis_time(redis_key, start_time)
```

Comment:
I think I was expecting some like:
```python
if message['helper_metadata'][self.fieldname].get('session_start') is not None:
    ...
```
rather than the Redis check? I think Redis should always be updated if we set something of the message because otherwise the two might get out of sync (apologies if my comments elsewhere suggested something different).